File: src/advent_of_code/y2022/day17.clj
(ns advent-of-code.y2022.day17
  (:require [advent-of-code.support :refer :all]
            [clojure.set :as set]
            [clojure.string :as str]))

(def input (get-input 2022 17))
(def tile-data "..####.\n\n...#...\n..###..\n...#...\n\n....#..\n....#..\n..###..\n\n..#....\n..#....\n..#....\n..#....\n\n..##...\n..##...")
(def tiles
  (->> (str/split tile-data #"\n\n")
       (mapv (fn [tile]
               (set (for [[y line] (indexed (reverse (str/split-lines tile)))
                          [x c] (indexed line)
                          :when (= c \#)]
                      [x y]))))))

(defn get-tile [i top]
  (set (for [[x y] (tiles i)] [x (+ y top 4)])))

(defn move [tile dx dy]
  (set (for [[x y] tile] [(+ x dx) (+ y dy)])))

(defn move-left [tile]
  (if (not-any? (comp #{0} first) tile)
    (move tile -1 0)
    tile))

(defn move-right [tile]
  (if (not-any? (comp #{6} first) tile)
    (move tile 1 0)
    tile))

(defn move-down [tile]
  (move tile 0 -1))

(defn try-move [tile move-fn screen]
  (let [tile' (move-fn tile)]
    (if (intersects? tile' screen)
      tile
      tile')))

(defn debug [screen]
  (doseq [y (take 12 (range (inc (apply max (map second screen))) -1 -1))]
    (println (->> (range 7)
                  (map #(if (contains? screen [% y]) "#" "."))
                  (apply str))))
  (println))

(defn fingerprint [screen top]
  (->> (range 7)
       (mapv (fn [x]
               (->> (range (- top 5) (inc top))
                    (transduce (comp (filter (fn [y] (screen [x y])))
                                     (map #(- top %)))
                               min
                               255))))
       (reduce (fn [res x]
                 (bit-or (bit-shift-left res 8) x))
               0)))

(defn solve [input rounds]
  (let [moves (str/trim input)]
    (loop [i 0
           move-i 0
           top 0
           skipped 0
           screen (set (for [x (range 7)] [x 0]))
           seen {}]
      (if (>= i rounds)
        (+ top skipped)
        (let [tile-i (mod i (count tiles))
              [move-i tile] (loop [move-i move-i
                                   tile (get-tile tile-i top)]
                              (let [move (nth moves move-i)
                                    move-i (mod (+ move-i 1) (count moves))
                                    tile (case move
                                           \< (try-move tile move-left screen)
                                           \> (try-move tile move-right screen))
                                    tile' (move-down tile)]
                                (if (intersects? tile' screen)
                                  [move-i tile]
                                  (recur move-i tile'))))
              top (apply max top (map second tile))
              screen (set/union screen tile)
              sig [tile-i move-i (fingerprint screen top)]]
          (if-let [[prev-i prev-top] (get seen sig)]
            (let [cycle (- i prev-i)
                  n (quot (- rounds i) cycle)
                  skipped (+ skipped (* n (- top prev-top)))
                  new-i (+ i 1 (* n cycle))]
              (recur new-i move-i top skipped screen seen))
            (recur (inc i) move-i top skipped screen (assoc seen sig [i top]))))))))

(defn solve1 [input]
  (solve input 2022))

(defn solve2 [input]
  (solve input 1000000000000))

File: src/advent_of_code/y2022/day17.py
from itertools import islice

MOVES = open("resources/inputs/2022/day17.txt").read().strip()
RAW_TILES = """\
..####.

...#...
..###..
...#...

....#..
....#..
..###..

..#....
..#....
..#....
..#....

..##...
..##..."""
TILES = [[int(line.replace(".", "0").replace("#", "1"), 2)
          for line in tile.splitlines()][::-1]
         for tile in RAW_TILES.split("\n\n")]


def debug(screen, prefix=""):
    start = True
    for line in reversed(screen):
        if start and not line:
            continue
        start = False
        print(prefix + bin(line)[2:].rjust(7, ".").replace("0", ".").replace("1", "#"))
    print()


def collision(screen, tile, y):
    return any(screen[y + i] & t for i, t in enumerate(tile))


def fingerprint(screen):
    res = 0
    for line in islice(filter(None, reversed(screen)), 10):
        res = (res << 7) | res | line
    return res


def solve(rounds):
    screen = [~0]
    top = 0
    skipped = 0
    seen = {}
    i = 0
    move_idx = 0
    while i < rounds:
        tile_idx = i % len(TILES)
        tile = TILES[tile_idx]

        while len(screen) < top + 4 + len(tile):
            screen.append(0)
        tile_y = top + 4

        while True:
            move = MOVES[move_idx]
            move_idx = (move_idx + 1) % len(MOVES)
            if move == "<":
                if not any(t & 1 << 6 for t in tile):
                    tile1 = [t << 1 for t in tile]
                    if not collision(screen, tile1, tile_y):
                        tile = tile1
            else:
                if not any(t & 1 for t in tile):
                    tile1 = [t >> 1 for t in tile]
                    if not collision(screen, tile1, tile_y):
                        tile = tile1

            if collision(screen, tile, tile_y - 1):
                break
            else:
                tile_y -= 1

        for ti, t in enumerate(tile):
            screen[tile_y + ti] |= t
        top = max(top, tile_y + len(tile) - 1)

        sig = (tile_idx, move_idx, fingerprint(screen))
        if prev := seen.get(sig):
            cycle = i - prev[0]
            n = (rounds - i) // cycle
            if n > 0:
                i += n * cycle
                skipped += n * (top - prev[1])
        else:
            seen[sig] = (i, top)
        i += 1

    # debug(screen[-20:])
    return top + skipped


part1 = solve(2022)
print(part1)
assert part1 == 3202

part2 = solve(1000000000000)
print(part2)
assert part2 == 1591977077352
